include/input_mappings.hpp

```diff
@@ -15,8 +18,79 @@ struct InputMappings {
 	}
 
 	void setMapping(Scancode scancode, u32 key) { container[scancode] = key; }
+
+	void serialize(const std::filesystem::path& path) const {
+		toml::basic_value<toml::preserve_comments, std::map> data;
+
+		std::error_code error;
+		if (std::filesystem::exists(path, error)) {
+			try {
+				data = toml::parse<toml::preserve_comments, std::map>(path);
+			} catch (const std::exception& ex) {
+				Helpers::warn("Exception trying to parse mappings file. Exception: %s\n", ex.what());
+				return;
+			}
+		} else {
+			if (error) {
+				Helpers::warn("Filesystem error accessing %s (error: %s)\n", path.string().c_str(), error.message().c_str());
+			}
+			printf("Saving new mappings file %s\n", path.string().c_str());
+		}
+
+		data["Metadata"]["Name"] = name.empty() ? "Unnamed Mappings" : name;
+		data["Metadata"]["Device"] = device.empty() ? "Unknown Device" : device;
+
+		data["Mappings"] = toml::table{};
+
+		for (const auto& [scancode, key] : container) {
+			if (data["Mappings"].contains(HID::Keys::keyToName(key)) == false) {
+				data["Mappings"][HID::Keys::keyToName(key)] = toml::array{};
+			}
+
+			data["Mappings"][HID::Keys::keyToName(key)].push_back(scancodeToName(scancode));
+		}
+
+		std::ofstream file(path, std::ios::out);
+		file << data;
+	}
+
+	void deserialize(const std::filesystem::path& path) {
+		toml::basic_value<toml::preserve_comments, std::map> data;
+
+		std::error_code error;
+		if (!std::filesystem::exists(path, error)) {
+			if (error) {
+				Helpers::warn("Filesystem error accessing %s (error: %s)\n", path.string().c_str(), error.message().c_str());
+			}
+			
+			return;
+		}
+
+		try {
+			data = toml::parse<toml::preserve_comments, std::map>(path);
+		} catch (const std::exception& ex) {
+			Helpers::warn("Exception trying to parse config file. Exception: %s\n", ex.what());
+			
+			return;
+		}
+
+		name = toml::find_or<std::string>(data, "Metadata.Name", "Unnamed Mappings");
```

Comment:
Is the Metadata.Name syntax legal? Can't find it in the docs.

Reply:
It is (dotted keys) but it doesn't work like I think it did